pact/dao/dao.pact

```diff
@@ -0,0 +1,279 @@
+        (enforce (> (now) (add-time lst-deactivate DEACTIVATE_COOLDOWN)) "Deactivate Cooldown Failure")
+        (update state DAO_STATE_KEY {"last-ambassador-deactivation":(now)})
+        (update ambassadors ambassador {"active":false}))
+      (with-capability (INTERNAL)
+        (adjust-ambassador-count -1)))
+      true)
+  (defun reactivate-ambassador:bool (guardian:string ambassador:string)
+    (with-capability (GUARDIAN guardian)
+      (update ambassadors ambassador {"active":true})
+      (with-capability (INTERNAL)
+        (adjust-ambassador-count 1)))
+      true)
+
+  ; ----
+  ; Freeze the DAO
+  (defun vote-to-freeze:bool (ambassador:string)
+    (with-capability (AMBASSADOR ambassador)
+      (update ambassadors ambassador {"voted-to-freeze":(now)}))
+    true)
+
+  (defun freeze:string (ambassador:string)
+    (with-capability (AMBASSADOR ambassador)
+    (let* ((live-ambs
+              (map (at "voted-to-freeze")
+              (select ambassadors ["voted-to-freeze"] (where 'active (= true)))))
+           (ambs-cnt (length live-ambs))
+           (ambs-voted-to-freeze (length (filter (btwn-incl (add-time (now) (- APPROVAL_COOLDOWN)) (now)) live-ambs)))
+           (dao-frz-ts (add-time (now) FREEZE_TIMEOUT)))
+      (enforce (> (* 2 ambs-voted-to-freeze) ambs-cnt)
+        (format "Majority vote failed: {} of {}" [ambs-voted-to-freeze ambs-cnt]))
+      (is-dao-frozen) ; so it can't get called twice
+      ; is this correct?
+      (update state DAO_STATE_KEY
+        {"dao-frozen-until": dao-frz-ts})
+      (format "DAO now frozen until {}" [dao-frz-ts]))))
+
+
+  ; ----
+  ; Upgrade the DAO
+  (defcap GOVERNANCE ()
+    (if (try false (require-capability (INTERNAL)))
+      true
+      (let ((x (is-dao-frozen)); if it's frozen, bail
+            (res (check-hash-approval (tx-hash))))
+      res)))
+
+  (defun check-hash-approval:bool (hsh:string)
+    (with-read state DAO_STATE_KEY
+      {'proposed-upgrade-time:=prp-time
+      ,'guardian-count:=grd-cnt}
+      (enforce (>= (now) (add-time prp-time APPROVAL_COOLDOWN))
+        (format "Proposal still in cooldown" [(add-time prp-time APPROVAL_COOLDOWN)]))
+      (enforce (< (now) (add-time prp-time APPROVAL_TIMEOUT))
+        (format "Proposal has timed out" [(add-time prp-time APPROVAL_COOLDOWN)]))
+      (let*
+         ((approvals (map (at 'approved-date )
+           (select guardians (where 'approved-hash (= hsh)))))
+         (valid-aprv-start (add-time (now) (- APPROVAL_TIMEOUT)))
+         (valid-approvals (length (filter (<= valid-aprv-start) approvals))))
+        (enforce (>= (* 2 valid-approvals) grd-cnt)
+          (format "Upgrade not approved, {} of {} before {}"
+            [valid-approvals approvals valid-aprv-start])))))
+
+    ; ----
+    ; Initialize the DAO
+    (defun init:string ()
+      (init-state)
+      (coin.create-account
+        DAO_ACCT_NAME
+        (create-module-guard DAO_ACCT_NAME)))
+
+    (defun state-hash:string ()
```

Comment:
What is this for? Also will not typecheck. If a hash is really needed consider storing it in state table.

Reply:
Mostly a utility function for testing. The idea that it gives the DAO a test that fails if the internal state changes (even when all the tests pass) so there isn't an accidental mutation to state. Thoughts? I can make it un-runable outside of the repl.

Reply:
It seems too opaque to function as a state regression. By itself, it could just as easily be regressing a bug or invariant violation. The only way to know it's correct is to verify the entire state row-by-row ... after which I suppose the hash acts as a golden record, but is maybe duplicative.